appeals/web/src/server/app/components/file-downloader.component.js

```diff
@@ -198,3 +197,106 @@ export const getDocumentDownloadByVersion = async ({ apiClient, params, session
 
 	return response.status(200);
 };
+
+/**
+ *
+ * @param {import('@pins/blob-storage-client').BlobStorageClient} blobStorageClient
+ * @param {string} blobStorageContainer
+ * @param {string} blobStoragePath
+ * @returns {Promise<*>}
+ */
+export const createBlobDownloadStream = async (
+	blobStorageClient,
+	blobStorageContainer,
+	blobStoragePath
+) => {
+	// Document URIs are persisted with a prepended slash, but this slash is treated as part of the key by blob storage so we need to remove it
+	const documentKey = blobStoragePath.startsWith('/') ? blobStoragePath.slice(1) : blobStoragePath;
+
+	const blobProperties = await blobStorageClient.getBlobProperties(
+		blobStorageContainer,
+		documentKey
+	);
+
+	if (!blobProperties) {
+		// skip download
+		return;
+	}
+
+	const blobDownloadResponseParsed = await blobStorageClient.downloadStream(
+		blobStorageContainer,
+		documentKey
+	);
+
+	if (!blobDownloadResponseParsed?.readableStreamBody) {
+		throw new Error(`Document ${documentKey} missing stream body`);
+	}
+
+	// @ts-ignore
+	return blobDownloadResponseParsed.blobDownloadStream;
+};
+
+/**
+ * Build a zipped file and Download
+ *
+ * @param {{apiClient: import('got').Got, params: {caseId: string, filename?: string}, session: SessionWithAuth}} request
+ * @param {Response} response
+ * @returns {Promise<Response>}
+ */
+export const getBulkDocumentDownload = async ({ apiClient, params, session }, response) => {
+	const { filename: requestedFilename, caseId } = params;
+
+	// console.log(JSON.stringify({ getBulkDocumentDownload1: { caseId, requestedFilename } }, null, 2));
+
+	const folders = await getFolders(apiClient, caseId);
+
+	const bulkFileInfo = folders
+		?.filter((folder) => folder.documents.length)
+		.flatMap((folder) => {
+			return folder.documents.map((document) => {
+				const { blobStorageContainer, blobStoragePath, documentURI } =
+					document.latestDocumentVersion;
+				return {
+					fullName: `${folder.path}/${document.name}`,
+					blobStorageContainer,
+					blobStoragePath,
+					documentURI
+				};
+			});
+		});
+
+	// Tell the browser that this is a zip file.
+	response.setHeader('content-type', 'application/zip');
+	response.setHeader('content-disposition', `attachment; filename=${requestedFilename}`);
+
+	// Create archive.
+	const archive = archiver('zip', {
+		zlib: { level: 9 } // Sets the compression level.
+	});
+
+	archive.pipe(response);
+
+	const blobStorageClient = await createBlobStorageClient(session);
+
+	if (!bulkFileInfo || !bulkFileInfo.length) {
+		return response.status(404);
+	}
+
+	const blobStreams = await Promise.all(
```

Comment:
There's a small chance a document record may not have the corresponding blob (especially on dev machines): if that happens, the stream gets locked... maybe append to the archive only if the blob stream is valid?

Reply:
I'm filtering out the empty blobStreams when appending to the archive to fix this issue.

Reply:
If there are any missing files, they will be reported within the missing-files.json file at the root of the zipped file